### DISCUSSION TITLE
[v5] Extend database dialects

### DISCUSSION THREAD

**Discussion by maxgalbu** [adonisjs/core#2995] [Help]
Hi, 
I'm trying to do some low level stuff on some fields, before and after database queries, and in v4 [I was able to extend dialects to implement my requirements](https://github.com/adonisjs/core/discussions/2324).

Is there an easy way to do the same in v5? Can I extend one of `MysqlDialect`,`QueryClient`, `DatabaseQueryBuilder`, and how does this is going to be imported in config/database.ts?

As background, I can't do it in beforeFetch because then I would need to do stuff also on ALL where() clauses and also auth related queries (hidden inside the framework)

**Comment by thetutlage** [comment]
Hey!

Here are the docs on extending the query builder https://docs.adonisjs.com/guides/database/query-builder#extending-query-builders. The MysqlDialect is part of knex codebase and I am not sure if they share any hooks for extending it.

Can you please in simple terms what exactly are you trying to achieve?

**Reply by maxgalbu** [reply]
I need to encrypt some fields when inserting/updating in db, and decrypt them when selecting from db. I need to call an API to do that so I need to do async stuff in nodejs + I can't do it from database.

I noticed that querybuilder's macros are now async... maybe that and beforeFetch() and beforeFind() can fit my needs.

**Reply by RomainLanz** [reply]
You can use the `consume` and `prepare` decorator's option for that.

```ts
@column({
  consume: (value?: Buffer) => (!value ? null : Encryption.decrypt(value.toString())),
  prepare: (value?: unknown) => (!value ? null : Encryption.encrypt(value)),
})
public secret: Record<string, string>
```

🔗 https://twitter.com/romainlanz/status/1283365113119019008
📚 https://docs.adonisjs.com/reference/orm/decorators#column

**Reply by maxgalbu** [reply]
@RomainLanz this could work but I need to call an API, so I would need to do a call *for each* field and *for each* query. This could slow down my app and the encrypt API

**Reply by RomainLanz** [reply]
Can you give more context? It seems that you would be better creating a dedicated service/action than using Lucid's hooks to do this kind of stuff.

**Reply by maxgalbu** [reply]
> Can you give more context?

I need to expose a login API - user can login and register using social or their own email address. Some of their info (email, name, surname, social provider id, etc) needs to be encrypted in db (don't mention it, legal stuff), so I need to encrypt/decrypt at runtime.

Adonis auth is obviously involved... even without using models, I would need to use Database and override stuff on where() and insert() to use adonis auth. And I don't want to reinvent the wheel of authentication.

**Reply by thetutlage** [reply]
As @RomainLanz said, you will be better off creating an abstraction layer and handle all this encryption inside it. 

As a side note: You can also opt for a database with encrypted disk. I think AWS offers that

**Reply by maxgalbu** [reply]
@thetutlage do you have an example or something that can show me the direction? How can I use this abstraction layer with adonisjs/auth? Do i need to extend auth?

Eg do you mean something like this? Here I'm extracting things from auth since I can't directly use models or database
```js
import AnonymizeService from "@ioc:My/EncryptService" 

...

async login() {
    //Manual login authentication
    const user = await User.query().where(await AnonymizeService.encryptFields({
        email: email,
    })).first();
    if (!user) {
        throw new InvalidInputException("Invalid email or password");
    }
    
    //Manual verify password
    const pwdVerify = await Hash.verify(password, user.password);
    if (!pwdVerify) {
        throw new InvalidInputException("Invalid email or password");
    }

    const userDecrypted = AnonymizeService.decryptFields(user);

    //Generate api token etc
}

async register() {
    //Validate etc

    const user = await User.create(await AnonymizeService.encryptFields({
        name: request.input("name"),
        surname: request.input("surname"),
        email: request.input("email"),
        password: request.input("password"),
    }));

    const userDecrypted = AnonymizeService.decryptFields(user); //or keep data from request?
    //Send emails etc
}
```

I was hoping

**Reply by RomainLanz** [reply]
Your code will not work since it should output a different value when you encrypt the same thing multiple times.
In your example, you use weak encryption since you can query on the email field with the encrypted value.

Concerning the usage of `@adonisjs/auth`, you are not obligated to use the `auth.attempt()` method to log a user in.

📚 https://docs.adonisjs.com/guides/auth/api-tokens-guard#authgenerate
📚 https://docs.adonisjs.com/guides/auth/web-guard#authlogin

**Reply by maxgalbu** [reply]
> In your example, you use weak encryption since you can query on the email field with the encrypted value.

Yes, it's a legal decision, weak encryption, no nonce, salt. That's what they want.

> Concerning the usage of @adonisjs/auth, you are not obligated to use the auth.attempt() method to log a user in.

ok! Then I shall use that example and just call generate() when needed. Let's see if it works

**Reply by maxgalbu** [reply]
Now I remember why i needed a low level solution. 

Since this needs to be done outside Model or Database (that is, like in the example I've posted before, with AnonymizeService), the model always reflects what's in the db. So, if I read a model from db, it will have its field encrypted, eg `User.email` is encrypted. If I need to work on these fields, i would need to decrypt in AND THEN store them somewhere on the model to avoid having to decrypt them again every time. 

This means having a property `unencryptedFields` on all models...  Maybe like this?
```
@column({
  consume: (value?: Buffer) => {
     return this.unencryptedFields.email ? this.unencryptedFields.email : value;
  }
})
public email: string

private unencryptedFields: UserUnencryptedFields = {}
```

Doesn't look good and it's a lot of work, but at least it hides the anonymization stuff outside Model... is there a better solution?

Having the encryption/decryption at low level (in the database class) could avoid all this, that's why I went for that in the v4 version

**Reply by mastermunj** [reply]
We also need to implement similar encryption very soon and have been thinking to do it using consume / prepare as explained by Romain.

Since we also have requirement to be able to search in database for the encrypted fields, the encryption will be weak, but nonetheless, we have enabled encryption at RDS level.

This is one of the reasons I requested https://github.com/adonisjs/lucid/issues/682.

As of now I am working on existsEncrypted and uniqueEncrypted validators of my own that helps me solve the purpose of encryption and search both through the model.
